In [20]:
# coding: utf8

# Введение в язык Python
## 3.1 Динамическая типизация или все есть объект.


In [5]:
x = 1
x = 'Hello'
y = x
x = 1.2
x = int
x = print

name = 'Ivan'
print(name)
name.upper()

Ivan


'IVAN'

In [14]:
id(x)

4336668960

In [6]:
id('hello')

4379319664

In [8]:
id('hello')

4379319664

In [6]:
x = 'he'+'llo'
id(x)

4380410416

In [8]:
id(1)

4337418480

In [10]:
id(2)

4337418512

In [11]:
id(1+1)

4337418512

## 3.2 Комментарии
**Комментарии** – это то, что пишется после символа #, и представляет интерес лишь как заметка для читающего программу.

Текст программы говорит о том, **КАК**, а комментарии должны объяснять, **ПОЧЕМУ**.

In [21]:
'''
print('Привет, Мир!') # print -- это функция
print() // 1
print() // 1
'''

# print('Привет, Мир!') # print -- это функция
# print() // 1
# print() // 1

print(comment_varible)


print('Привет, Мир!') # print -- это функция
print() // 1
print() // 1



## 3.3 Литеральные константы
Примеры литеральных констант: **5, 1.23, 9.25e-3, "It's a string!"**.

In [42]:
x = 5 * 1.23 * 0.5 * .5
x = "It's a string"
x = 'Hello World!'

MINUTES_IN_HOUR = 60
minutes_in_hour = 24

print(MINUTES_IN_HOUR, minutes_in_hour)


60 24


In [49]:
value = 'fdgdgf'
print(value, id(value))
print(value.upper(), id(value.upper()))

fdgdgf 4415887280
FDGDGF 4419940720


In [58]:
int_value_1 = 5
int_value_2 = 2
print(id(int_value_1))

int_value_1 = int_value_1 + int_value_2
print(id(int_value_1))

int_value_1 = 0
print(id(int_value_1))

4337418608
4337418672
4337418448


In [65]:
int_value_1 = 7
int_value_2 = 5
print(int_value_2, id(int_value_2))

int_value_3 = int_value_1 - 2
print(int_value_3, id(int_value_3))

5 4337418608
5 4337418608


In [66]:
int_value_1 = 1
int_value_2 = 100500
print(int_value_2, id(int_value_2))

int_value_3 = 100502 - 2 
print(int_value_3, id(int_value_3))

100500 4414031152
100500 4414037488


In [67]:
int_value_1 = 0
int_value_2 = 5
print(int_value_2, id(int_value_2))
int_value_3 = int_value_1 + int_value_2
print(int_value_3, id(int_value_2), id(int_value_3))

5 4337418608
5 4337418608 4337418608


## 3.3 Имена идентификаторов
Переменные – это частный случай идентификаторов. Идентификаторы – это имена, присвоенные чему-то для его обозначения.

При выборе имён для идентификаторов необходимо соблюдать следующие правила:

1) первым символом идентификатора должна быть буква из алфавита (символ ASCII в верхнем или нижнем регистре, или символ Unicode), а также символ подчёркивания (“_”);

2) остальная часть идентификатора может состоять из букв(символы ASCII вверхнем или нижнем регистре, а также символы Unicode), знаков подчёркивания (“_”) или цифр (0-9);

3) имена идентификаторов чувствительны к регистру.

Примеры допустимых имён идентификаторов: i,__my_name,name_23,a1b2_c3, δξѪђёўЩӆΞέά.

Примеры недопустимых имён идентификаторов: 2things, my- name, >a1b2_c3 и "это_в_кавычках".

## 3.3 Типы данных

<table style="width: 500px;">
 <thead style="background: #e7e7e7; font-weight: bold;">
    <tr >
        <td style="text-align: center;">Неизменяемые</td>
        <td style="text-align: center;">Изменяемые</td>
    </tr>
  </thead>
   <tbody >
    <tr>
        <td style="text-align: center;">str</td>
        <td style="text-align: center;">list</td>
    </tr>
    <tr>
        <td style="text-align: center;">int</td>
        <td style="text-align: center;">set</td>
    </tr>
    <tr>
        <td style="text-align: center;">float</td>
        <td style="text-align: center;">dict</td>
    </tr>
    <tr>
        <td style="text-align: center;">long</td>
        <td style="text-align: center;"></td>
    </tr>
    <tr>
        <td style="text-align: center;">bool</td>
        <td style="text-align: center;"></td>
    </tr>
    <tr>
        <td style="text-align: center;">tuple</td>
        <td style="text-align: center;"></td>
    </tr>
    <tr>
        <td style="text-align: center;">frozenset</td>
        <td style="text-align: center;"></td>
    </tr>
  </tbody>
</table>


### 3.3.1 Числа
**Числа** в Python бывают трёх типов.

1. Целые числа, например: **2, 31, 100500**
2. Числа с плавающей точкой, например: **3.23, 52.3E-4, 0.5, .05**
3. Комплексные числа, например: **(-5 + 4j), (2.3 - 4.6j)**

In [68]:
x = 1_00500
x = .05
x = (-5 + 4j)
x = 2 + 3j
print(x)
x = 10001000
x = 1_001_000
x

(2+3j)


1001000

### 3.3.2 Строки
**Строка** – это последовательность символов. Чаще всего строки – это просто некоторые наборы слов.
Слова могут быть как на английском языке, так и на любом другом, поддерживаемом стандартом Unicode, что означает почти на любом языке мира.

In [4]:
x = 'Hello World'
x = "Hello World"

# Символ - это срока, состоящая из одного элемента
x = 'a'
x = "a"

# Объединение строк
x = "Hello " "World"
print(x)
x = 'Hello ' + 'World'
print(x)

# Использование специальных символов
x = "I'm writing cool code."
x = 'I\'m writing cool code.'
print(x)

# Многострочный способ
x = """
    Beautiful is better than ugly.
    Explicit is better than implicit.
    """
print(x)
x = 'Beautiful is better than ugly.\n' + \
    'Explicit is better than implicit.'

x = 'Beautiful is better than ugly.\n' \
    'Explicit is better than implicit.'

x = (
    'Beautiful is better than ugly.\n' 
    'Explicit is better than implicit.')

x = 'Beautiful is better than ugly.\n\
Explicit is better than implicit.'
print(x)

Hello World
Hello World
I'm writing cool code.

    Beautiful is better than ugly.
    Explicit is better than implicit.
    
Beautiful is better than ugly.
Explicit is better than implicit.


#### Форматирование строк
https://www.python.org/dev/peps/pep-3101/

In [18]:
name = 'Ibrahim'
age = '49'

# Format
print("I'm {age} years old {name}".format(age=age, name=name))
print("I'm {0} years old {1}, Dear {1}, bla".format(age, name))
print("I'm {} years old {}, Dear {}, bla".format(age, name, name))
print("My name is {0} :-{{}}".format(name))

# Prefered format
print(f"I'm {age} years old {name}")
print(f"I'm age={age} years old name={name}")
print(f"I'm {age=} years old {name=}")


# Alternative
print("I'm %s" % name)
print("I'm %s years old %s" % (age, name))

# Numbers
print('\nNumbers')
print('{0:.3}'.format(1/3))
print('{0:+}'.format(100500))
print('{0:=10}'.format(100500))
print('{0:=010}'.format(100500))
print('{0:=+010}'.format(100500))
print('{0:g}'.format(100111111111))
print('{0:,}'.format(10011111199999111))
print('{0:.2f}'.format(1/3))
print('{:010.4f}'.format(1/3))
print('{:%}'.format(1/3))
print('{:.1%}'.format(1/3))

# More things
print('\nMore things')
print('{0:_^11}'.format(name))
print('{0:^11}'.format(name))
print('{0:_<10}'.format(name))
print('{0:_>10}'.format(name))

I'm 49 years old Ibrahim
I'm 49 years old Ibrahim, Dear Ibrahim, bla
I'm 49 years old Ibrahim, Dear Ibrahim, bla
My name is Ibrahim :-{}
I'm 49 years old Ibrahim
I'm age=49 years old name=Ibrahim
I'm age='49' years old name='Ibrahim'
I'm Ibrahim
I'm 49 years old Ibrahim

Numbers
0.333
+100500
    100500
0000100500
+000100500
1.00111e+11
10,011,111,199,999,111
0.33
00000.3333
33.333333%
33.3%

More things
__Ibrahim__
  Ibrahim  
Ibrahim___
___Ibrahim


## 3.4 Многострочный код

In [21]:
x = 10
print(x)

# Альтернативы
x = 10;
print(x);

x = 10; print(x)

print\
(x)
print('Veryveryveryveryveryveryveryveryveryveryveryvery ' \
      'long string')
print('Veryveryveryveryveryveryveryveryveryveryveryvery ' + 
      'long string')
print('Veryveryveryveryveryveryveryveryveryveryveryvery ' 
      'long string')

10
10
10
10
Veryveryveryveryveryveryveryveryveryveryveryvery long string
Veryveryveryveryveryveryveryveryveryveryveryvery long string
Veryveryveryveryveryveryveryveryveryveryveryvery long string


## Отступы: боль ли это?
Предложения, идущие вместе, должны иметь одинаковый отступ. Каждый такой набор предложений называется блоком.

**Не смешивайте пробелы и символы табуляции в отступах**. Рекомендуется использовать одиночную табуляцию или **четыре пробела** для каждого уровня отступа.

In [28]:
i = 5
 print('Значение составляет ', i) # Ошибка! Пробел в начале строки
print('Я повторяю, значение составляет ', i)

IndentationError: unexpected indent (312479578.py, line 2)

## 3.5 if __name__ == '__main__'

In [35]:
if __name__ == '__main__':
    print('Hello World!')

In [32]:
import math
print(math.__name__)

math


## 3.6 Задачи
### 1. Welcome to Python
 **Дано:** имя и фамилия.
 
 **Задание:** написать программу, которая будет приветствовать нового человека в мире Python.
     Текст приветсвия: Hello **NAME SURNAME**! You just delved into Python. Great start!
     
 **Пример:**
     Hello Ibrahim Petrov! You just delved into Python. Great start!
 
 ### 2. Python art
 **Дано:** маркер (символ) и толщина фигуры.
 
 **Задание:** написать программу, которая будет отображать заданную фигуру.
 
 **Пример:** Маркер = **H**, толщина **5**.
                        

In [ ]:
    H    
   HHH   
  HHHHH  
 HHHHHHH 
HHHHHHHHH
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHHHHHHHHHHHHHHHHHHHHHH   
  HHHHHHHHHHHHHHHHHHHHHHHHH   
  HHHHHHHHHHHHHHHHHHHHHHHHH   
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHH               HHHHH             
                    HHHHHHHHH 
                     HHHHHHH  
                      HHHHH   
                       HHH    
                        H 

In [ ]:
thickness = 5
c = 'H'

#Top Cone
for i in range(thickness):
    print((c*i).______(thickness-1) + c + (c*i).______(thickness-1))

#Top Pillars
for i in range(thickness+1):
    print((c*thickness).______(thickness*2) + (c*thickness).______(thickness*6))

#Middle Belt
for i in range((thickness+1)//2):
    print((c*thickness*5).______(thickness*6))    

#Bottom Pillars
for i in range(thickness+1):
    print((c*thickness).______(thickness*2) + (c*thickness).______(thickness*6))    

#Bottom Cone
for i in range(thickness):
    print(((c*(thickness-i-1)).______(thickness) + c + (c*(thickness-i-1)).______(thickness)).______(thickness*6))  

 ### 3. Заголовок
 **Дано:** текст любой длины.
 
 **Задание:** написать программу, которая выведет заголовок, используя заданный текст.
 Подсказка используйте метод **title**.
 
 **Пример:** text = 'hello world'; результат = Hello World
 
 ### 4. Форматированный вывод денежной суммы
 **Дано:** денежная сумма (amount > 0).
 
 **Задание:** написать программу, которая распечатает число в принятом денежном формате **XXX,XXX.XX**.
 
 
 **Пример:** amount = 100500.157; результат = 100,500.16
 
 ### [Junior] 5. Дизайнер ковриков
 Дизайнер составил шаблон домашних ковриков. Для массового выпуска ковриков ему нужно уметь быстро составлять макет произвольного размера. Известно, что длина коврика всегда больше в 3 раза чем его ширина (**W = 3 * H**).
 
 **Дано:** ширина коврика.
 
 **Задание:** написать программу, которая будет составлять макет коврика для его дальнейшего производства.
 
 **Примеры:** 

In [ ]:
    Size: 7 x 21 
    ---------.|.---------
    ------.|..|..|.------
    ---.|..|..|..|..|.---
    -------WELCOME-------
    ---.|..|..|..|..|.---
    ------.|..|..|.------
    ---------.|.---------
    
    Size: 11 x 33
    ---------------.|.---------------
    ------------.|..|..|.------------
    ---------.|..|..|..|..|.---------
    ------.|..|..|..|..|..|..|.------
    ---.|..|..|..|..|..|..|..|..|.---
    -------------WELCOME-------------
    ---.|..|..|..|..|..|..|..|..|.---
    ------.|..|..|..|..|..|..|.------
    ---------.|..|..|..|..|.---------
    ------------.|..|..|.------------
    ---------------.|.---------------

 ### [Junior+] 6. Произведение цифр
 **Дано:** целое число.
 
 **Задание:** написать программу, которая перемножит все цифры заданного числа (0 - исключить).
 
 
 **Примеры:** 
 1) value = 123405, результат = 120
 2) value = 999, результат = 729
 3) value = 1000, результат = 1
 4) value = 1111, результат = 1

### Решение задания №2

In [ ]:
thickness = 5
c = 'H'

#Top Cone
for i in range(thickness):
    print((c*i).rjust(thickness-1) + c + (c*i).ljust(thickness-1))

#Top Pillars
for i in range(thickness+1):
    print((c*thickness).center(thickness*2) + (c*thickness).center(thickness*6))

#Middle Belt
for i in range((thickness+1)//2):
    print((c*thickness*5).center(thickness*6))    

#Bottom Pillars
for i in range(thickness+1):
    print((c*thickness).center(thickness*2) + (c*thickness).center(thickness*6))    

#Bottom Cone
for i in range(thickness):
    print(((c*(thickness-i-1)).rjust(thickness) + c + (c*(thickness-i-1)).ljust(thickness)).rjust(thickness*6))  

### Решение задания №5

In [ ]:
height = 11
width = height * 3

for stick_count in range(1, height, 2):
    print(('.|.' * stick_count).center(width, '-'))

print('WELCOME'.center(width, '-'))

for stick_count in range(height-2, 0, -2):
    print(('.|.' * stick_count).center(width, '-'))
